src/main/java/org/kiwiproject/registry/consul/server/ConsulRegistryService.java

```diff
@@ -0,0 +1,220 @@
+import static org.kiwiproject.registry.util.Ports.findFirstPortPreferSecure;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.orbitz.consul.Consul;
+import com.orbitz.consul.model.agent.ImmutableRegCheck;
+import com.orbitz.consul.model.agent.ImmutableRegistration;
+import com.orbitz.consul.model.agent.Registration;
+import lombok.extern.slf4j.Slf4j;
+import org.kiwiproject.base.KiwiEnvironment;
+import org.kiwiproject.base.Optionals;
+import org.kiwiproject.base.UUIDs;
+import org.kiwiproject.registry.config.ServiceInfo;
+import org.kiwiproject.registry.consul.config.ConsulRegistrationConfig;
+import org.kiwiproject.registry.exception.RegistrationException;
+import org.kiwiproject.registry.model.Port;
+import org.kiwiproject.registry.model.ServiceInstance;
+import org.kiwiproject.registry.server.RegistryService;
+import org.kiwiproject.registry.util.Paths;
+import org.kiwiproject.retry.SimpleRetryer;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * {@link RegistryService} implementation for Consul discovery servers
+ * <p>
+ * NOTE: This class is intended to manage the registration and unregistration of a service with Consul.  If your service
+ * is using a framework that has Consul integration, you might not want to use this class as it may attempt to register
+ * more than once (once from here and once from the framework integration)
+ */
+@Slf4j
+public class ConsulRegistryService implements RegistryService {
+
+    /**
+     * Maximum number of attempts we will make trying to register with Eureka (via POST)
+     */
+    @VisibleForTesting
+    static final int MAX_REGISTRATION_ATTEMPTS = 60;
+
+    /**
+     * Value for delay to wait if a call to Eureka fails.
+     */
+    private static final long RETRY_DELAY = 1;
+
+    /**
+     * Unit for delay to wait if a call to Eureka fails.
+     */
+    private static final TimeUnit RETRY_DELAY_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * Value for delay when un-registering if a call to Eureka fails.
+     */
+    private static final long UNREGISTER_RETRY_DELAY = 3;
+
+    /**
+     * Time unit for delay when un-registering if a call to Eureka fails.
+     */
+    private static final TimeUnit UNREGISTER_RETRY_DELAY_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * Maximum number of attempts we will make to un-register from Eureka (via DELETE request).
+     */
+    private static final int MAX_UNREGISTER_ATTEMPTS = 5;
+
+    private final Consul consul;
+    private final ConsulRegistrationConfig config;
+    private final SimpleRetryer registerRetryer;
+    private final SimpleRetryer unregisterRetryer;
+
+    @VisibleForTesting
+    final AtomicReference<ServiceInstance> registeredService;
+
+    public ConsulRegistryService(Consul consul, ConsulRegistrationConfig config, KiwiEnvironment environment) {
```

Comment:
Should we have a constructor without the `KiwiEnvironment` that calls this one with a `new DefaultEnvironment()`? What do we do for the Eureka service?